torch/_subclasses/fake_tensor.py

```diff
@@ -87,6 +99,39 @@ def __call__(self, t, device=None):
             return self.from_meta_and_device(t, device)
 
 
+def run_cpu_fallback(func, args, kwargs, orig_not_implemented_exception):
+    with no_dispatch():
+        def to_cpu(e):
+            if isinstance(e, FakeTensor):
+                return torch.empty_like(e, device="cpu")
+            return e
+        try:
+            args = tree_map(to_cpu, args)
+            kwargs = tree_map(to_cpu, kwargs)
+            r = func(*args , **kwargs)
+        except Exception:
+            # original error more orinformative
+            raise orig_not_implemented_exception
+        tensor_impls = set()
+
+        def collect_impls(e):
+            if isinstance(e, torch.Tensor):
+                tensor_impls.add(e)
+
+        tree_map(collect_impls, (args, kwargs))
+        # proper aliasing/metadata relationship between outputs and inputs will
+        # not be set up, bc of conversion to cpu, error on reused impls
+
+        def throw_on_reused_impls(e):
+            if e in tensor_impls:
+                raise orig_not_implemented_exception
+
+        tree_map(throw_on_reused_impls, r)
```

Comment:
This seems very insufficient and also the case being tested for here feels like it can be handled.

It's insufficient in that this fallback will not work for view operations, and you won't capture errors here because view operators will return a fresh tensor with shared storage. Testing for shared storage is better.

It's handleable in that if an output of the cpu operation is exactly the same as the input, we know we can just return the original meta in that case and it will be OK.

Reply:
Yes. But you can also have an operator that doesn't return the input but does mutate the metadata and you have no way of detecting this. Well, tags would help.